File: src/formatters/markdown.ts
// Markdown output formatter

import type { MessageOutput, ProcessedChannel, ProcessedMessage } from '../types'
import { formatDateLong, formatRelativeTime, formatTime } from '../utils/date'

export interface MarkdownOptions {
  relative?: boolean
}

export function formatMarkdown(outputs: MessageOutput[], options: MarkdownOptions = {}): string {
  const sections: string[] = []

  for (const output of outputs) {
    sections.push(formatChannelMarkdown(output, options.relative ?? false))
  }

  return sections.join('\n\n---\n\n')
}

function channelHeaderMarkdown(channel: ProcessedChannel): string {
  if (channel.type === 'dm') {
    return `## DMs with ${escapeMarkdown(channel.name)}`
  }
  if (channel.type === 'group') {
    return `## Group DM: ${escapeMarkdown(channel.name)}`
  }
  const display = channel.displayName ? ` (${escapeMarkdown(channel.displayName)})` : ''
  return `## #${escapeMarkdown(channel.name)}${display}`
}

function formatChannelMarkdown(output: MessageOutput, relative: boolean): string {
  const { channel, messages } = output
  const lines: string[] = []

  lines.push(channelHeaderMarkdown(channel))
  lines.push('')

  // Group messages by date
  const messagesByDate = groupByDate(messages)

  for (const [date, msgs] of messagesByDate) {
    lines.push(`### ${date}`)
    lines.push('')

    for (const msg of msgs) {
      lines.push(formatMessage(msg, relative))
      lines.push('')
    }
  }

  // Add redaction summary if any
  if (output.redactions.length > 0) {
    lines.push('')
    lines.push(`_${output.redactions.length} secret(s) redacted_`)
  }

  const state = output.retrieval.selection.queryTruncated
  lines.push('')
  lines.push(
    `_Coverage: ${output.retrieval.selection.selectedCount} selected, ${output.retrieval.visiblePostCount} visible; query ${state === true ? 'truncated' : state === false ? 'complete' : 'completeness unknown'}_`,
  )
  if (output.retrieval.selection.nextCursor) {
    lines.push(`Next cursor: \`${output.retrieval.selection.nextCursor}\``)
  }

  return lines.join('\n')
}

function formatMessage(msg: ProcessedMessage, relative: boolean, depth: number = 0): string {
  const timeStr = relative ? formatRelativeTime(msg.timestamp) : formatTime(msg.timestamp)
  const lines: string[] = []
  const prefix = depth > 0 ? '> '.repeat(depth) : ''

  const safePermalink = safeHttpUrl(msg.permalink)
  const postId = escapeMarkdown(msg.id)
  const postRef = safePermalink ? `[${postId}](<${safePermalink}>)` : postId
  const markers = messageMarkers(msg)
  lines.push(
    `${prefix}**${escapeMarkdown(msg.user)}**${markers ? ` ${markers}` : ''} (${timeStr}, ${postRef}):`,
  )

  // Quote the message content
  const quotePrefix = `${prefix}> `
  lines.push(quoteLines(escapeMarkdown(msg.text), quotePrefix))
  const stateParts = [`Updated ${msg.updatedAt.toISOString()}`]
  if (msg.editedAt) stateParts.push(`edited ${msg.editedAt.toISOString()}`)
  if (msg.deletedAt) stateParts.push(`deleted ${msg.deletedAt.toISOString()}`)
  lines.push(`${quotePrefix}_${stateParts.join('; ')}_`)

  // Add file attachments if any
  if (msg.fileDetails.length > 0) {
    const files = msg.fileDetails
      .map((file) => {
        const label = file.name ? `${file.name} (${file.id})` : file.id
        const details = [
          file.mime,
          file.extension,
          file.size === undefined ? undefined : `${file.size} B`,
        ]
          .filter(Boolean)
          .join(', ')
        return escapeMarkdown(`${label}${details ? `, ${details}` : ''}`)
      })
      .join(', ')
    lines.push(`${quotePrefix}_Files: ${files}_`)
  }

  for (const attachment of msg.attachments) {
    const title = attachment.title ?? 'Attachment'
    const safeTitleLink = attachment.titleLink ? safeHttpUrl(attachment.titleLink) : undefined
    if (attachment.pretext) lines.push(quoteLines(escapeMarkdown(attachment.pretext), quotePrefix))
    lines.push(
      safeTitleLink
        ? `${quotePrefix}**[${escapeMarkdown(title)}](<${safeTitleLink}>)**`
        : `${quotePrefix}**${escapeMarkdown(title)}**`,
    )
    if (attachment.authorName)
      lines.push(quoteLines(`By: ${escapeMarkdown(attachment.authorName)}`, quotePrefix))
    if (attachment.text) lines.push(quoteLines(escapeMarkdown(attachment.text), quotePrefix))
    for (const field of attachment.fields ?? []) {
      lines.push(
        quoteLines(
          `${field.title ? `**${escapeMarkdown(field.title)}:** ` : ''}${escapeMarkdown(field.value ?? '')}`,
          quotePrefix,
        ),
      )
    }
    if (attachment.fallback)
      lines.push(quoteLines(`Fallback: ${escapeMarkdown(attachment.fallback)}`, quotePrefix))
    if (attachment.footer)
      lines.push(quoteLines(`_${escapeMarkdown(attachment.footer)}_`, quotePrefix))
    if (attachment.color) lines.push(`${quotePrefix}Color: ${escapeMarkdown(attachment.color)}`)
    if (attachment.timestamp)
      lines.push(`${quotePrefix}Timestamp: ${escapeMarkdown(attachment.timestamp)}`)
    for (const url of [
      attachment.authorLink,
      attachment.authorIcon,
      attachment.footerIcon,
      attachment.imageUrl,
      attachment.thumbUrl,
    ]) {
      const safeUrl = url ? safeHttpUrl(url) : undefined
      if (safeUrl) lines.push(`${quotePrefix}<${safeUrl}>`)
    }
  }

  if (msg.reactions.length > 0) {
    const reactions = msg.reactions
      .map(({ emoji, count, actors }) => {
        const names = actors.map((actor) => escapeMarkdown(actor.username ?? actor.id)).join(', ')
        return `:${escapeMarkdown(emoji)}: ${count}${names ? ` (${names})` : ''}`
      })
      .join(' · ')
    lines.push(`${quotePrefix}_Reactions: ${reactions}_`)
  }

  // Render replies
  if (msg.replies && msg.replies.length > 0) {
    lines.push('')
    for (const reply of msg.replies) {
      lines.push(formatMessage(reply, relative, depth + 1))
      lines.push('')
    }
  }

  return lines.join('\n')
}

function messageMarkers(msg: ProcessedMessage): string {
  const markers: string[] = []
  if (msg.isDeleted) markers.push('[deleted]')
  else if (msg.editedAt) markers.push('[edited]')
  if (msg.isSystem)
    markers.push(msg.postType ? `[system:${escapeMarkdown(msg.postType)}]` : '[system]')
  if (msg.isPinned) markers.push('[pinned]')
  return markers.join(' ')
}

function escapeMarkdown(value: string): string {
  return value.replace(/([\\`*_[\]{}()<>#+\-.!|~])/g, '\\$1')
}

function quoteLines(value: string, prefix: string): string {
  return value
    .split('\n')
    .map((line) => `${prefix}${line}`)
    .join('\n')
}

function safeHttpUrl(url: string): string | undefined {
  try {
    const parsed = new URL(url)
    if (parsed.protocol !== 'http:' && parsed.protocol !== 'https:') return undefined
    return url.replace(/</g, '%3C').replace(/>/g, '%3E').replace(/\\/g, '%5C')
  } catch {
    return undefined
  }
}

function groupByDate(messages: ProcessedMessage[]): Map<string, ProcessedMessage[]> {
  const groups = new Map<string, ProcessedMessage[]>()

  // Sort messages oldest first for display
  const sorted = [...messages].sort((a, b) => a.timestamp.getTime() - b.timestamp.getTime())

  for (const msg of sorted) {
    const date = formatDateLong(msg.timestamp)
    if (!groups.has(date)) {
      groups.set(date, [])
    }
    groups.get(date)?.push(msg)
  }

  return groups
}

File: internal/output/markdown.go
package output

import (
	"fmt"
	"io"
	"net/url"
	"slices"
	"strings"
)

type MarkdownOptions struct {
	Relative bool
}

// FormatMarkdown renders already-presented values without further sanitizing them.
func FormatMarkdown(outputs []MessageOutput, dates DateFormatter, options MarkdownOptions) string {
	sections := make([]string, 0, len(outputs))
	for _, output := range outputs {
		sections = append(sections, formatMarkdownSection(output, dates, options.Relative))
	}
	return strings.Join(sections, "\n\n---\n\n")
}

// WriteMarkdown renders and writes exactly once. It never retries an uncertain write.
func WriteMarkdown(w io.Writer, outputs []MessageOutput, dates DateFormatter, options MarkdownOptions) (int, error) {
	formatted := FormatMarkdown(outputs, dates, options)
	n, err := io.WriteString(w, formatted)
	if err == nil && n != len(formatted) {
		err = io.ErrShortWrite
	}
	return n, err
}

func formatMarkdownSection(output MessageOutput, dates DateFormatter, relative bool) string {
	lines := []string{markdownChannelHeader(output.Channel), ""}
	messages := slices.Clone(output.Messages)
	slices.SortStableFunc(messages, func(a, b Message) int { return a.Timestamp.Compare(b.Timestamp) })
	lastDate := ""
	for _, message := range messages {
		date := dates.FormatDateLong(message.Timestamp)
		if date != lastDate {
			lines = append(lines, "### "+date, "")
			lastDate = date
		}
		lines = append(lines, formatMarkdownMessage(message, dates, relative, 0), "")
	}
	if len(output.Redactions) > 0 {
		lines = append(lines, "", fmt.Sprintf("_%d secret(s) redacted_", len(output.Redactions)))
	}
	state := "completeness unknown"
	if output.Retrieval.Selection.QueryTruncated != nil {
		if *output.Retrieval.Selection.QueryTruncated {
			state = "truncated"
		} else {
			state = "complete"
		}
	}
	lines = append(lines, "", fmt.Sprintf("_Coverage: %d selected, %d visible; query %s_", output.Retrieval.Selection.SelectedCount, output.Retrieval.VisiblePostCount, state))
	if cursor := output.Retrieval.Selection.NextCursor; cursor != nil && *cursor != "" {
		lines = append(lines, "Next cursor: `"+*cursor+"`")
	}
	return strings.Join(lines, "\n")
}

func markdownChannelHeader(channel Channel) string {
	switch channel.Type {
	case "unknown":
		return "## Unknown channel (" + escapeMarkdown(channel.ID) + ")"
	case "dm":
		return "## DMs with " + escapeMarkdown(channel.Name)
	case "group":
		return "## Group DM: " + escapeMarkdown(channel.Name)
	default:
		display := ""
		if channel.DisplayName != "" {
			display = " (" + escapeMarkdown(channel.DisplayName) + ")"
		}
		return "## #" + escapeMarkdown(channel.Name) + display
	}
}

func formatMarkdownMessage(message Message, dates DateFormatter, relative bool, depth int) string {
	timeText := dates.FormatTime(message.Timestamp)
	if relative {
		timeText = dates.FormatRelativeTime(message.Timestamp)
	}
	prefix := strings.Repeat("> ", depth)
	postID := escapeMarkdown(message.ID)
	postRef := postID
	if permalink, ok := safeHTTPURL(message.Permalink); ok {
		postRef = "[" + postID + "](<" + permalink + ">)"
	}
	markers := markdownMessageMarkers(message)
	header := prefix + "**" + escapeMarkdown(message.User) + "**"
	if markers != "" {
		header += " " + markers
	}
	header += " (" + timeText + ", " + postRef + "):"
	quotePrefix := prefix + "> "
	lines := []string{header, markdownQuoteLines(escapeMarkdown(message.Text), quotePrefix)}
	lines = append(lines, quotePrefix+"_"+formatStateTimes(message)+"_")
	if len(message.FileDetails) > 0 {
		lines = append(lines, quotePrefix+"_Files: "+escapeMarkdown(formatFiles(message))+"_")
	}
	for _, attachment := range message.Attachments {
		title := attachment.Title
		if title == "" {
			title = "Attachment"
		}
		if attachment.Pretext != "" {
			lines = append(lines, markdownQuoteLines(escapeMarkdown(attachment.Pretext), quotePrefix))
		}
		var titleLine string
		if titleLink, ok := safeHTTPURL(attachment.TitleLink); ok {
			titleLine = "**[" + escapeMarkdown(title) + "](<" + titleLink + ">)**"
		} else {
			titleLine = "**" + escapeMarkdown(title) + "**"
		}
		lines = append(lines, markdownQuoteLines(titleLine, quotePrefix))
		if attachment.AuthorName != "" {
			lines = append(lines, markdownQuoteLines("By: "+escapeMarkdown(attachment.AuthorName), quotePrefix))
		}
		if attachment.Text != "" {
			lines = append(lines, markdownQuoteLines(escapeMarkdown(attachment.Text), quotePrefix))
		}
		for _, field := range attachment.Fields {
			value := escapeMarkdown(field.Value)
			if field.Title != "" {
				value = "**" + escapeMarkdown(field.Title) + ":** " + value
			}
			lines = append(lines, markdownQuoteLines(value, quotePrefix))
		}
		if attachment.Fallback != "" {
			lines = append(lines, markdownQuoteLines("Fallback: "+escapeMarkdown(attachment.Fallback), quotePrefix))
		}
		if attachment.Footer != "" {
			lines = append(lines, markdownQuoteLines("_"+escapeMarkdown(attachment.Footer)+"_", quotePrefix))
		}
		if attachment.Color != "" {
			lines = append(lines, quotePrefix+"Color: "+escapeMarkdown(attachment.Color))
		}
		if attachment.Timestamp != "" {
			lines = append(lines, quotePrefix+"Timestamp: "+escapeMarkdown(attachment.Timestamp))
		}
		for _, candidate := range []string{attachment.AuthorLink, attachment.AuthorIcon, attachment.FooterIcon, attachment.ImageURL, attachment.ThumbURL} {
			if safe, ok := safeHTTPURL(candidate); ok {
				lines = append(lines, quotePrefix+"<"+safe+">")
			}
		}
	}
	if len(message.Reactions) > 0 {
		reactions := make([]string, 0, len(message.Reactions))
		for _, reaction := range message.Reactions {
			actors := make([]string, 0, len(reaction.Actors))
			for _, actor := range reaction.Actors {
				name := actor.Username
				if name == "" {
					name = actor.ID
				}
				actors = append(actors, escapeMarkdown(name))
			}
			value := fmt.Sprintf(":%s: %d", escapeMarkdown(reaction.Emoji), reaction.Count)
			if len(actors) > 0 {
				value += " (" + strings.Join(actors, ", ") + ")"
			}
			reactions = append(reactions, value)
		}
		lines = append(lines, quotePrefix+"_Reactions: "+strings.Join(reactions, " · ")+"_")
	}
	if len(message.Replies) > 0 {
		lines = append(lines, "")
		for _, reply := range message.Replies {
			lines = append(lines, formatMarkdownMessage(reply, dates, relative, depth+1), "")
		}
	}
	return strings.Join(lines, "\n")
}

func markdownMessageMarkers(message Message) string {
	markers := make([]string, 0, 3)
	if message.IsDeleted {
		markers = append(markers, "[deleted]")
	} else if message.EditedAt != nil {
		markers = append(markers, "[edited]")
	}
	if message.IsSystem {
		if message.PostType != "" {
			markers = append(markers, "[system:"+escapeMarkdown(message.PostType)+"]")
		} else {
			markers = append(markers, "[system]")
		}
	}
	if message.IsPinned {
		markers = append(markers, "[pinned]")
	}
	return strings.Join(markers, " ")
}

var markdownEscaper = strings.NewReplacer(
	"\\", "\\\\", "`", "\\`", "*", "\\*", "_", "\\_", "[", "\\[", "]", "\\]",
	"{", "\\{", "}", "\\}", "(", "\\(", ")", "\\)", "<", "\\<", ">", "\\>",
	"#", "\\#", "+", "\\+", "-", "\\-", ".", "\\.", "!", "\\!", "|", "\\|", "~", "\\~",
)

func escapeMarkdown(value string) string { return markdownEscaper.Replace(value) }

func markdownQuoteLines(value, prefix string) string {
	return prefix + strings.ReplaceAll(value, "\n", "\n"+prefix)
}

func safeHTTPURL(value string) (string, bool) {
	if value == "" {
		return "", false
	}
	parsed, err := url.Parse(value)
	if err != nil || parsed.Host == "" || (parsed.Scheme != "http" && parsed.Scheme != "https") {
		return "", false
	}
	replacer := strings.NewReplacer("<", "%3C", ">", "%3E", "\\", "%5C")
	return replacer.Replace(value), true
}
